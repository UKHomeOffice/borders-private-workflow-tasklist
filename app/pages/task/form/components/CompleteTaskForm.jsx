import React from 'react';
import PropTypes from 'prop-types';
import { customEventSubmissionStatus, form, loadingTaskForm, submissionStatus } from '../selectors';
import { bindActionCreators } from 'redux';
import * as taskFormActions from '../actions';
import * as taskActions from '../../display/actions';
import {Formio} from 'react-formio';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import AppConstants from '../../../../common/AppConstants';
import NotFound from '../../../../core/components/NotFound';
import { FAILED, SUBMISSION_SUCCESSFUL, SUBMITTING } from '../constants';
import TaskForm from './TaskForm';
import withLog from '../../../../core/error/component/withLog';
import DataSpinner from '../../../../core/components/DataSpinner';
import { unclaimSuccessful } from '../../display/selectors';

export class CompleteTaskForm extends React.Component {
  constructor(props) {
    super(props);
    this.handleCustomEvent = this.handleCustomEvent.bind(this);
    this.handleSubmission = this.handleSubmission.bind(this);
  }

  componentDidMount() {
    this.props.fetchTaskForm(this.props.task);
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    const { submissionStatus, customEventSubmissionStatus} = this.props;
    if (this.form && this.form.formio) {
      if (submissionStatus !== prevProps.submissionStatus) {
        this.handleSubmission(submissionStatus);
      }
      if (customEventSubmissionStatus !== prevProps.customEventSubmissionStatus
          && customEventSubmissionStatus === SUBMISSION_SUCCESSFUL) {
        this.props.history.replace(AppConstants.YOUR_TASKS_PATH);
      }
      if (this.props.unclaimSuccessful) {
        this.props.history.push(AppConstants.YOUR_TASKS_PATH);
      }
    }
  }

  handleSubmission(submissionStatus) {
    const taskName = this.props.task.get('name');
    const path = this.props.history.location.pathname;
    const user = this.props.kc.tokenParsed.email;
    switch (submissionStatus) {
      case SUBMITTING :
        this.props.log([{
          user: user,
          path: path,
          level: 'info',
          message: `Submitting data for ${taskName}`
        }]);
        break;
      case SUBMISSION_SUCCESSFUL:
        this.form.formio.emit('submitDone');
        this.props.log([{
          user: user,
          path: path,
          level: 'info',
          message: `${taskName} successfully completed`
        }]);
        Formio.clearCache();
        if (this.props.redirectPath) {
          this.props.history.replace(this.props.redirectPath);
        } else {
          this.props.history.replace(AppConstants.YOUR_TASKS_PATH);
        }
        break;
      case FAILED :
        this.props.log([{
          user: user,
          path: path,
          level: 'error',
          message: `Failed to complete ${taskName}`
        }]);
        this.form.formio.emit('error');
        this.form.formio.emit('change', this.form.formio.submission);
        break;
      default:
        this.props.log([{
          level: 'warn',
          path: path,
          user: user,
          message: 'Unknown submission status',
          submissionStatus: submissionStatus
        }]);
    }
  }

  handleCustomEvent = (event, variableName) => {
    const { task } = this.props;
    if (event.type === 'unclaim') {
      this.props.unclaimTask(task.get('id'));
    } else if (event.type === 'cancel') {
      this.props.history.replace(AppConstants.YOUR_TASKS_PATH);
    } else {
      this.props.customEvent(event, task, variableName);
    }
  };

  componentWillUnmount() {
    this.form = null;
    this.props.resetForm();
  }


  render() {
    const { loadingTaskForm, form, task} = this.props;
    if (loadingTaskForm) {
      return <DataSpinner message="Loading form for task..."/>;
    }
    if (!form) {
      return <NotFound resource="Form" id={task.get('name')}/>;
    }
    return <TaskForm {...this.props}
                onSubmitTaskForm={(submissionData, variableName) => {
                  this.props.submitTaskForm(form._id, task.get('id'),
                    submissionData, variableName);
                }}
                onCustomEvent={(event) => this.handleCustomEvent(event)}
                formReference={(form) => this.form = form}/>
  }
}

CompleteTaskForm.propTypes = {
  submitTaskForm: PropTypes.func.isRequired,
  customEvent: PropTypes.func.isRequired,
  fetchTaskForm: PropTypes.func.isRequired,
  unclaimTask: PropTypes.func.isRequired,
  resetForm: PropTypes.func.isRequired,
  loadingTaskForm: PropTypes.bool,
  submissionStatus: PropTypes.string,
  customEventSubmissionStatus: PropTypes.string,
};


const mapDispatchToProps = dispatch => bindActionCreators(Object.assign({}, taskFormActions, taskActions), dispatch);


export default withRouter(connect((state) => {
  return {
    form: form(state),
    loadingTaskForm: loadingTaskForm(state),
    unclaimSuccessful: unclaimSuccessful(state),
    submissionStatus: submissionStatus(state),
    customEventSubmissionStatus: customEventSubmissionStatus(state),
    kc: state.keycloak
  };
}, mapDispatchToProps)(withLog(CompleteTaskForm)));
