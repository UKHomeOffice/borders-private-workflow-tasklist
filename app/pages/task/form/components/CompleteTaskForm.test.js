import React from 'react';
import { mount} from 'enzyme';
import configureStore from 'redux-mock-store';
import Immutable from 'immutable';
import {CompleteTaskForm} from './CompleteTaskForm';

const { Map } = Immutable;


const taskForm = {
  "type": "form",
  "tags": [
    "common"
  ],
  "owner": "XXXXX",
  "components": [
    {
      "autofocus": false,
      "id": "id",
      "input": true,
      "tableView": true,
      "inputType": "text",
      "inputMask": "",
      "label": "Text",
      "key": "text",
      "placeholder": "",
      "prefix": "",
      "suffix": "",
      "multiple": false,
      "defaultValue": "",
      "protected": false,
      "unique": false,
      "persistent": true,
      "hidden": false,
      "clearOnHide": true,
      "spellcheck": true,
      "validate": {
        "required": false,
        "minLength": "",
        "maxLength": "",
        "pattern": "",
        "custom": "",
        "customPrivate": false
      },
      "conditional": {
        "show": "",
        "when": null,
        "eq": ""
      },
      "type": "textfield",
      "labelPosition": "top",
      "tags": [],
      "properties": {}
    },
    {
      "id": "canceloperation",
      "autofocus": false,
      "input": true,
      "label": "Cancel operation",
      "tableView": false,
      "key": "canceloperation",
      "size": "md",
      "leftIcon": "",
      "rightIcon": "",
      "block": false,
      "action": "event",
      "disableOnInvalid": false,
      "theme": "primary",
      "type": "button",
      "tags": [],
      "conditional": {
        "show": "",
        "when": null,
        "eq": ""
      },
      "properties": {
        "success-message": "Operation successfully cancelled"
      },
      "event": "cancel-operation"
    }
  ],
  "_id": "XXXXX",
  "display": "form",
  "submissionAccess": [
    {
      "roles": [
        "5b0fa1b2769993003d6fba6f"
      ],
      "type": "create_own"
    },
    {
      "roles": [
        "XXXXX"
      ],
      "type": "read_own"
    },
    {
      "roles": [
        "XXXX"
      ],
      "type": "update_own"
    },
    {
      "roles": [
        "XXXXX"
      ],
      "type": "delete_own"
    }
  ],
  "title": "Custom Button Event",
  "name": "customButtonEvent",
  "path": "custombuttonevent",
  "access": [
    {
      "roles": [
        "XXXX",
        "XXXX",
        "XXXXX"
      ],
      "type": "read_all"
    }
  ],
  "created": "2019-02-01T07:13:35.002Z",
  "modified": "2019-02-01T07:13:35.069Z",
  "machineName": "customButtonEvent"
};
describe('CompleteTaskForm Component', () => {
  const initialState = {
    'task-form': new Map({
      loadingTaskForm: false,
      form: null,
      submittingToFormIO: false,
      submissionToFormIOSuccessful: false,
      submittingTaskFormForCompletion: false,
      taskFormCompleteSuccessful: null,
      customEventSuccessfullyExecuted: false,
      submittingCustomEvent: false
    })
  };
  const mockStore = configureStore();
  let store;
  beforeEach(() => {
    store = mockStore(initialState);
  });
  const fetchTaskForm = jest.fn();
  const resetForm = jest.fn();
  const submitTaskForm = jest.fn();
  const unclaimTask = jest.fn();
  const customEvent = jest.fn();

  it('renders loading text', async() => {
    const props = {
      formName: 'testForm',
      loadingTaskForm : true,
      task: Immutable.fromJS({
        assignee: null,
        name: 'testTask'
      })
    };
    const wrapper = await mount(
      <CompleteTaskForm store={store} {...props}
                        fetchTaskForm={fetchTaskForm}
                        resetForm={resetForm}
                        submitTaskForm={submitTaskForm}
                        unclaimTask={unclaimTask}
                        customEvent={customEvent}
      />
    );
    console.log(wrapper.debug());
    expect(fetchTaskForm).toBeCalled();
    expect(wrapper.name()).toEqual('CompleteTaskForm');
    expect(wrapper.html()).toEqual('<div id="dataSpinner"><div class="loader-content"><div color="black" style="color: black;" class="sk-fade-in sk-spinner line-spin-fade-loader"><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div></div></div><div class="loader-message"><strong class="govuk-!-font-weight-bold">Loading form for task...</strong></div></div>');
    wrapper.unmount();
    expect(resetForm).toBeCalled();
  });

  it('renders task form', async() =>{
    const props = {
      formName: 'testForm',
      loadingTaskForm: false,
      form: taskForm,
      task: Immutable.fromJS({
        assignee: null,
        name: 'testTask'
      }),
      variables: {

      }
    };
    const wrapper = await shallow(
      <CompleteTaskForm store={store} {...props}
                        fetchTaskForm={fetchTaskForm}
                        resetForm={resetForm}
                        submitTaskForm={submitTaskForm}
                        unclaimTask={unclaimTask}
                        customEvent={customEvent}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

});
