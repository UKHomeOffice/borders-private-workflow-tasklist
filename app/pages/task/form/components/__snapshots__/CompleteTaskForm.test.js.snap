// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompleteTaskForm Component renders task form 1`] = `
<Loader
  backgroundStyle={
    Object {
      "backgroundColor": "white",
    }
  }
  foregroundStyle={
    Object {
      "color": "black",
    }
  }
  hideContentOnLoad={false}
  message={
    <div
      style={
        Object {
          "borderRadius": 0,
          "justifyContent": "center",
          "left": 0,
          "margin": "auto",
          "position": "absolute",
          "right": 0,
          "top": "-20px",
          "zIndex": 100,
        }
      }
    >
      <DataSpinner
        message="Submitting data..."
      />
    </div>
  }
  priority={0}
  show={false}
>
  <TaskForm
    customEvent={[MockFunction]}
    fetchTaskForm={
      [MockFunction] {
        "calls": Array [
          Array [
            Immutable.Map {
              "assignee": null,
              "name": "testTask",
            },
          ],
          Array [
            Immutable.Map {
              "assignee": null,
              "name": "testTask",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    form={
      Object {
        "access": Array [
          Object {
            "roles": Array [
              "XXXX",
              "XXXX",
              "XXXXX",
            ],
            "type": "read_all",
          },
        ],
        "components": Array [
          Object {
            "autofocus": false,
            "clearOnHide": true,
            "conditional": Object {
              "eq": "",
              "show": "",
              "when": null,
            },
            "defaultValue": "",
            "hidden": false,
            "id": "id",
            "input": true,
            "inputMask": "",
            "inputType": "text",
            "key": "text",
            "label": "Text",
            "labelPosition": "top",
            "multiple": false,
            "persistent": true,
            "placeholder": "",
            "prefix": "",
            "properties": Object {},
            "protected": false,
            "spellcheck": true,
            "suffix": "",
            "tableView": true,
            "tags": Array [],
            "type": "textfield",
            "unique": false,
            "validate": Object {
              "custom": "",
              "customPrivate": false,
              "maxLength": "",
              "minLength": "",
              "pattern": "",
              "required": false,
            },
          },
          Object {
            "action": "event",
            "autofocus": false,
            "block": false,
            "conditional": Object {
              "eq": "",
              "show": "",
              "when": null,
            },
            "disableOnInvalid": false,
            "event": "cancel-operation",
            "id": "canceloperation",
            "input": true,
            "key": "canceloperation",
            "label": "Cancel operation",
            "leftIcon": "",
            "properties": Object {
              "success-message": "Operation successfully cancelled",
            },
            "rightIcon": "",
            "size": "md",
            "tableView": false,
            "tags": Array [],
            "theme": "primary",
            "type": "button",
          },
        ],
        "created": "2019-02-01T07:13:35.002Z",
        "display": "form",
        "id": "XXXXX",
        "machineName": "customButtonEvent",
        "modified": "2019-02-01T07:13:35.069Z",
        "name": "customButtonEvent",
        "owner": "XXXXX",
        "path": "custombuttonevent",
        "submissionAccess": Array [
          Object {
            "roles": Array [
              "5b0fa1b2769993003d6fba6f",
            ],
            "type": "create_own",
          },
          Object {
            "roles": Array [
              "XXXXX",
            ],
            "type": "read_own",
          },
          Object {
            "roles": Array [
              "XXXX",
            ],
            "type": "update_own",
          },
          Object {
            "roles": Array [
              "XXXXX",
            ],
            "type": "delete_own",
          },
        ],
        "tags": Array [
          "common",
        ],
        "title": "Custom Button Event",
        "type": "form",
      }
    }
    formName="testForm"
    formReference={[Function]}
    loadingTaskForm={false}
    onCustomEvent={[Function]}
    onSubmitTaskForm={[Function]}
    resetForm={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    submitTaskForm={[MockFunction]}
    task={
      Immutable.Map {
        "assignee": null,
        "name": "testTask",
      }
    }
    unclaimTask={[MockFunction]}
    variables={Object {}}
  />
</Loader>
`;
